### ISSUE TITLE
[EKS] [request]: Add/Delete/Update Subnets Registered with the Control Plane

### ISSUE THREAD

**Issue by christopherhein** [aws/containers-roadmap#170]
**Tell us about your request**
The ability to update the Subnets that the EKS Control plane is registered with.

**Which service(s) is this request for?**
EKS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
https://twitter.com/julien_fabre/status/1099071498621411329

**Are you currently working around this issue?**

**Additional context**

**Comment by christopherhein** [comment]
@Pryz

**Comment by jmeichle** [comment]
This would be a nice improvement

**Comment by Pryz** [comment]
To add some color, here are some use cases : 

1.  You have a multi-tenant cluster configured with X number of subnets but you are getting close to IP exhaustion and you want to extend the setup to Y more subnets. Without losing the current configuration of course.

2. You are expending your setup to new availability zones and so want to use new subnets to schedule PODs there.

3. You were using your cluster on private subnets only and now want to extend to use some public subnets.

Generally, in many environment, network setup are moving and EKS need to be flexible enough to embrace such changes.

Thanks !

**Comment by devkid** [comment]
@Pryz Your worker nodes don't have to be in the same subnets that your control plan is configured for. The latter are used for creating the ENIs that are used for `kubectl log|exec|attach` and for ELB/NLB placement.

**Comment by Pryz** [comment]
@devkid yes but that's a problem. You can basically schedule PODs on subnets which are not configured on the master but can't access it (logs, proxy, whatever).

**Comment by devkid** [comment]
@Pryz If you have proper routing between the different subnet, this is not a problem. We have configured our control plane for one set of subnets and our workers run in a second, disjoint set of subnets and logs, proxy, exec are working just fine.

**Comment by hanjunlee** [comment]
@Pryz Could you explain detail how to routing between the different subnets? I am just wondering how to access to disjoint set of subnets. I hope your reply! Thanks

**Comment by Pryz** [comment]
@hanjunlee I'm not sure I understand your question. My setup is quite simple : 1 VPC, up to 5 CIDRs with 6 subnets each (3 private subnets, 3 public subnets).
Each AZ gets 2 routing tables (1 for private subnets and 1 for public subnets). There is no issue in this routing setup. Any IP from a subnet can talk with any other IP from any of the other subnets.

**Comment by aschonefeld** [comment]
I am seeing the same issue. I have configured 2 subnets initially but my CIDR range was too small for IP assignemts from the cluster. So I added new subnets to the VPC and the worker nodes are running fine in these new subnets.

When using kubectl proxy and accessing the URL I get an error of :

Error: 'Address is not allowed'
Trying to reach: 'https://<secondary-ip-of-worker-node>:8443/'

Control Pane ENI in old subnets, worker nodes in new subnets and kubectl host have all in and outbound rules for each other. I would think this is related to the issue of new subnets not having an attached ENI for the control pane. Any help would be appreciated.

**Comment by ckassen** [comment]
@aschonefeld did you tag the new subnets with the correct `kubernetes.io/cluster/<clustername>` tag set to `shared`?

I could not reproduce the error when I tagged the subnet that way. Spun up a node in the subnet and started a pod on that node. Afterwards I could run `kubectl logs|exec|port-forward` on that Pod.

**Comment by aschonefeld** [comment]
@ckassen the cloud formation temlate for the worker nodes tagged the subnet to shared. Kubectl logs for example is also working for me but proxy to the dashboard is not working. Command goes through but no connect to the dashboard is possible. Is kubectl. proxy working for you?

**Comment by willthames** [comment]
If you decide your load balancers are in the wrong subnet, and decide to create new ones, as far as I can tell EKS doesn't detect the new subnets, and still creates the load balancers in the old subnets, even though they're no longer tagged with `kubernetes.io/role/internal-elb`. Being able to add the new subnets to EKS would be useful.

**Comment by thomasjungblut** [comment]
Any work-around known so far? Terraform wants to create a new EKS cluster for me after adding new subnets :-/

**Comment by sjmiller609** [comment]
@thomasjungblut are you using this module? https://github.com/terraform-aws-modules/terraform-aws-eks the cluster aws_eks_cluster.this resource wants to be replaced?

**Comment by thomasjungblut** [comment]
We built our own module, but effectively it's the same TF resource that wants to be replaced yes.  

It makes sense, since the API doesn't support changing the subnets: https://docs.aws.amazon.com/eks/latest/APIReference/API_UpdateClusterConfig.html

Would be cool to at least have the option of adding new subnets. The background is that we want to switch from public to private subnets, so we could add our private subnets on top of the existing public ones and just change the routes a bit. Would certainly make our life a bit easier :)

**Comment by theothermike** [comment]
We just ran into this exact issue, using that EKS TF Module too.   A workaround that seems to work:

1) Create the new subnets, setup the routes, etc
2) Manually edit the ASG for the worker nodes, add the subnets
3) Edit the control plane SG, add the CIDR ranges of the new subnet

This, of course, breaks running TF with the EKS module for that cluster again.   We're hoping to mitigate that by tightening up the TF code so we can just create new properly sized VPC/subnets, and kill the old EKS cluster entirely

**Comment by jahzielHA** [comment]
We are having the same problem. We needed to add new subnets to the EKS cluster and we had to rebuild it since the `aws eks update-cluster-config --resources-vpc-config`  dose not allow to update Subnet nor Security Groups once the cluster has been built.

```
An error occurred (InvalidParameterException) when calling the UpdateClusterConfig operation: subnetIds and securityGroupIds cannot be updated.
```

**Comment by flythebluesky** [comment]
Do we really need to rebuild the entire cluster to add a subnet?

**Comment by mailjunze** [comment]
Workaround : If your EKS cluster is behind the current latest version and planning to upgrade.

Control-plane discovers the subnets at the time of initialization process. Tag the new subnets.  Upgrade the cluster, new tagged subnets will be discovered  automatically.

**Comment by henkka** [comment]
> Workaround : If your EKS cluster is behind the current latest version and planning to upgrade.
> 
> Control-plane discovers the subnets at the time of initialization process. Tag the new subnets. Upgrade the cluster, new tagged subnets will be discovered automatically.

This workaround didn't work for us :( We created new subnets, tagged them with `kubernetes.io/cluster/eks: shared` tag and ran the EKS upgrade with no changes to the EKS attached subnets. Anything we missed?

**Comment by QCU266** [comment]
> Workaround : If your EKS cluster is behind the current latest version and planning to upgrade.
> Control-plane discovers the subnets at the time of initialization process. Tag the new subnets. >
> Upgrade the cluster, new tagged subnets will be discovered automatically.

The same as @henkka , The workaround didn't work for me too.  
what should I do?

**Comment by BertieW** [comment]
@QCU Did you modify the control plane security groups with the new CIDRs?

**Comment by qcu** [comment]
> @qcu Did you modify the control plane security groups with the new CIDRs?

^-- @QCU266: this was for you.

**Comment by BertieW** [comment]
The process that worked for us using terraform:
Created new subnet(s)
Tagged new subnets with kubernetes.io/cluster/<clustername> tag set to shared -- our subnets share the same route table, but if they didn't, we'd have tagged that too.
Modify security group for the control plane to add the new CIDR.
Cluster schedules pods just fine, and stuff like logs|proxy|etc work with no issue.

We use the EKS terraform module, and all of this was duable with terraform. The worker node block will happily accept a subnet that isn't one of those declared with the cluster initially. No manual changes required.  This is all terraformed.

**Comment by khacminh** [comment]
Is there any plan for this proposal? The proposal is created for almost a year and AWS seems doesn't have any plan for it.

**Comment by cazter** [comment]
@BertieW using the TF module with the new subnets added to the worker groups "forces replacement" of the EKS cluster.

All subnets (w/ tags), are added to the VPC and SG's, and they share same route table and are within the same VPC. As @thomasjungblut pointed out, the TF module appears to be restricted by the AWS API limitations and cannot add the additional subnets to the existing cluster. I don't see anyway to get around replacing the EKS cluster if deploying from the TF module :(

**Comment by cazter** [comment]
Alright found a solution for the TF EKS module.

You can't make changes to the module subnets parameter. So if you were calling these subnets via a variable as I was like so:
> subnets             = module.vpc.private_subnets
You'll need to provide a static definition for precisely those subnets used when creating your cluster, example:
> subnets             = ["subnet-085cxxxx", "subnet-0a60xxxx", "subnet-0be8xxxx"]

Then within your worker groups you can add your new subnets. Afterwards you'll be able to apply the TF without forcing a cluster replacement.

**Comment by mailjunze** [comment]
> Workaround : If your EKS cluster is behind the current latest version and planning to upgrade.
> 
> Control-plane discovers the subnets at the time of initialization process. Tag the new subnets. Upgrade the cluster, new tagged subnets will be discovered automatically.

This workaround is not working anymore. The subnets chosen while creating the EKS cluster will be used by control plane nodes only. However, users can use different  subnets for workerNodes. EKS will be able to register worker nodes on different subnets which were not part of initial set of subnets when the cluster was created. 

You may attach a new **CIDR range** to the VPC and carve subnets out of newly created CIDR range and add tags to it to make sure EKS discovers those new subnets.

Worker nodes can be launched in new CIDR range subnets in the same VPC  after cluster creation without any issues.

**Comment by sbonasu** [comment]
+1

**Comment by chunsli** [comment]
+1

**Comment by jancurn** [comment]
+1

**Comment by shackit** [comment]
This has been one of those issues that should just be available IMHO, having to rebuild a cluster just to provide the cluster with some additional target subnets is really undesirable. We are in the position of having to rebuild one of our test clusters (luckily not a cluster hosting production workloads)

**Comment by blongman-snapdocs** [comment]
Been searching for a way to do this for a while.  

We have 6 different EKS clusters that were built with either only public or only private subnets.  I'd very much like to correct all 6 of them.  They were all built with terraform-aws-modules/eks/aws, and I'd like to keep maintaining them with such.

**Comment by thorro** [comment]
Another issue: unused subnet was deleted, but EKS still has it listed under Networking. Tried to turn on CloudWatch logging as we have problems with K8s API, but it fails with the message:

`InvalidRequestException
The subnet ID 'subnet-xxxx892f' does not exist (Service: AmazonEC2; Status Code: 400; Error Code: InvalidSubnetID.NotFound; Request ID: ce958c55-3c8a-4154-8cfa-25891bf8ca3f)`

**Comment by connatix-cradulescu** [comment]
We are facing same issue. We are trying to increase number of subnets on the EKS cluster an this is not possible without cluster recreation. This scenario have been tested about was support too with could formation and they reach same result I did with terraform.  The only solution is to change the core behavior of EKS to update number of subnets on the fly. We can't afford to redeploy an entire production cluster just because EKS does not support a "normal action" to update the numbers of subnets that are using. Please provide a feedback when this functionality will be implemented cause production clusters are affected by this missing.

**Comment by siarhei-karanets-epam** [comment]
I will try
https://aws.amazon.com/premiumsupport/knowledge-center/eks-multiple-cidr-ranges/

**Comment by manikandanmuthu44** [comment]
+1 this is very much needed is there any timeline on the addition of subnets without recreation ?

**Comment by thorro** [comment]
Just don't delete any of the subnets (it will let you if there are no used IPs), or you will be forced to recreate an entire EKS cluster.

**Comment by manikandanmuthu44** [comment]
@thorro So we cannot start new pods in the cluster since there are not enough IPs. So in the same VPC we have created some more subnets is it possible to use those subnets for solving this use-case without recreating the cluster ?

**Comment by thorro** [comment]
@manikandanmuthu44 we had no problems with that, just tag them.

**Comment by tomfotherby** [comment]
We also ran out of IP addresses due to initially choosing too smaller subnets for EKS. We created bigger subnets, tagged them, and untagged the smaller subnets, and one-by-one terminated the nodes waiting for each new node to accept pods before going onto the next one. It solved our problem in production with no downtime.

The only issue is the control plane is stuck with the old subnets, meaning we can't tidy them up.

**Comment by manikandanmuthu44** [comment]
@tomfotherby Thanks whatever you said worked out we did similar to it, created new bigger subnets and tagged them and booted nodes got associated with cluster and pods started picking up IPs in the new bigger subnet

**Comment by ab77** [comment]
The behaviour of not allowing subnets to be modified after creation is highly undesirable.

```
$ aws eks update-cluster-config --name eks-cluster --resources-vpc-config subnetIds=subnet-abc,subnet-def,subnet-123

An error occurred (InvalidParameterException) when calling the UpdateClusterConfig operation: subnetIds and securityGroupIds cannot be updated.
```

**Comment by fergaldonlon** [comment]
In my case a one character typo on terraform resulted in a cluster being created in the wrong AZ:
`availability_zone = var.azs[2]
`
I have tried a couple of things but in the end it was faster to rebuild the cluster. 
It would appear there is no easy way to edit the subnets on EKS.

**Comment by shikloshi** [comment]
+1

**Comment by nitrag** [comment]
+1 for EKS feature

I was able to add AZ/subnets manually but it was a hack to not be able to configure the EKS cluster the same.

**Comment by Ticiano-mw** [comment]
+1

@nitrag  how did you accomplish this? you were able to add subnets to an existing EKS cluster?

**Comment by nitrag** [comment]
@Ticiano-mw 

Added subnets, added route table association. Updated ASG's with additional vpc_zone_identifier's (subnets).

**Comment by sindhujaaitha07** [comment]
@Ticiano-mw how did you accomplish this?

**Comment by Ticiano-mw** [comment]
> @Ticiano-mw
> 
> Added subnets, added route table association. Updated ASG's with additional vpc_zone_identifier's (subnets).

But there are no ASGs for the EKS control plane hosts

**Comment by lfpalacios** [comment]
Does anyone has a workaround for this? None of the solutions presented worked for me...
Btw, this issue has a long time, AWS should address this asap

**Comment by fndiaz** [comment]
Unfortunately it is necessary to recreate the cluster to update the subnets.

This is very bad, I need to include a subnet and I will need to redo my 5 clusters

**Comment by dogzzdogzz** [comment]
We used to have /20 subnets but ip pool was not enough, we created extra subnets (/19), routing and ASG with EKS required tags, itworks very well, the new subnets doesn’t have to be registered to control plane.

**Comment by mf-lit** [comment]
I was in a similar position recently. A single worker group in private subnets without enough IPs in the CIDR.

FWIW here are the steps I took with the EKS module:

1. Added secondary VPC CIDR and three additional private subnets to use that range. Key at the stage is hard-coding the subnets IDs that are passed to the EKS module (as mentioned by **cazter** above) otherwise terraform will want to destroy and recreate the existing autoscaling group(s) which will destroy the worker nodes too.
2. Added a second worker group in the new subnets that were created in step 1. Again, the subnets IDs for these need to be hardcoded when passed to the EKS module.
3. Disable cluster-autoscaler on the original worker group. (This is done with the `autoscaling_enabled = false` option in the EKS module)
4. Gracefully drain all the nodes in the original worker group, cluster-autoscaler will bring up new nodes in the new group as needed and pods will migrate to them.
5. Scale the original group to zero (set `asg_min_size=0` and `asg_max_size=0`, apply and then delete the nodes manually)
6. At this point the hardcoded subnets can be swapped out for the appropriate resource attributes.

This doesn't change the fact that only the original subnets are registered with the control plane of course, but as others have said it does give fully operational pods and those pods are accessible by the control plane. Plus the terraform code is still sane when you're finished.

There is however a big gotcha, custom apiservices no longer work as the pods that back them are now in the new subnets and the control plane can't discover them. This breaks autoscaling for example because metrics-server is required.

```
kubectl get apiservice
NAME                                   SERVICE                                      AVAILABLE                      AGE
[...]
v1beta1.metrics.k8s.io                 kube-system/metrics-server                   False (FailedDiscoveryCheck)   104d
```

I worked around this by bringing back up the original workergroup (in the original subnets) and using taints/tolerations and nodeselectors to pin the only metrics-server pods to the workers in that group. Unfortunately bringing back that worker group also means going back to hardcoded subnet definitions. So it's a workable but very undesirable solution.

```
kubectl get apiservice
NAME                                   SERVICE                                      AVAILABLE                      AGE
[...]
v1beta1.metrics.k8s.io                 kube-system/metrics-server                   True                           104d
```

As everyone else has already said, this needs fixing properly.

**Comment by blacabexg** [comment]
+1

**Comment by savealive** [comment]
I can confirm that control plane restart has fixed "Address not allowed" issue. 
Assuming new subnets in secondary CIDR have have proper tags `kubernetes.io/cluster/mycluster,Value=shared`.
Enable custom CNI config in aws-node daemonset:
```
        - "name": "AWS_VPC_K8S_CNI_CUSTOM_NETWORK_CFG"
          "value": "true"
        - "name": "ENI_CONFIG_LABEL_DEF"
          "value": "failure-domain.beta.kubernetes.io/zone"
```
apply ENIconfig configmap:
```
apiVersion: crd.k8s.amazonaws.com/v1alpha1
kind: ENIConfig
metadata:
  name: us-east-1a
spec:
  securityGroups:
    - sg-0f50a320<id>
  subnet: subnet-03bfe23a<id>
...
# and similar for other AZs
```
Roll worker nodes and see that pods are now being assigned into new subnet (Nodes subnets remain "old" ones, no reason to change subnets in ASG).
```
metrics-server-b6866958f-q9fk2             1/1     Running   0          20h   10.7.13.140   ip-10-7-2-113.ec2.internal   <none>           <none>
metrics-server-b6866958f-w7cfq             1/1     Running   0          20h   10.7.10.132   ip-10-7-1-157.ec2.internal   <none>           <none>
```
Check API resources status and find it's failed `https://10.7.13.71:8443/apis/metrics.k8s.io/v1beta1: Address is not allowed`:
```
k describe apiservices.apiregistration.k8s.io v1beta1.metrics.k8s.io
...
    Message:               failing or missing response from https://10.7.13.71:8443/apis/metrics.k8s.io/v1beta1: Get https://10.7.13.71:8443/apis/metrics.k8s.io/v1beta1: Address is not allowed
    Reason:                FailedDiscoveryCheck
    Status:                False
    Type:                  Available
Events:                    <none>
```
After that simply update EKS version from version 1.17 to 1.18 and check status after updating:
```
k describe apiservices.apiregistration.k8s.io v1beta1.metrics.k8s.io
Name:         v1beta1.metrics.k8s.io
...
Status:
  Conditions:
    Last Transition Time:  2020-10-15T19:26:42Z
    Message:               all checks passed
    Reason:                Passed
    Status:                True
    Type:                  Available
Events:                    <none>
```

I suppose EKS control plan reads subnets tags on startup time and creates a list to pass to `--proxy-cidr-whitelist` parameter. Adding new subnets while cluster is running doesn't trigger API server config reload so new cidr are not in `--proxy-cidr-whitelist` list.
On the other hand version update causes EKS control plane to restart so it regenerates allowed IP ranges to connect/proxy by control plane.

This behavior has been tested on at least 3 different clusters.
Unfortunately there's no way to restart EKS control plan on demand so version upgrade is the only way to do it.

**Comment by ten-lac** [comment]
I just upgraded my EKS Managed Node Groups from 1.17 -> 1.18.  During the upgrade process, My node groups went degraded because one of the AZ no longer had instance type availability.  I could not change the node group network settings for this selection.  I tried to adjust this through the ASG directly which did not reflect on the control plane side.  This really needs to be fixed.  An upgrade process should not be this painful for a managed service.

**Comment by srfaytkn** [comment]
+1

**Comment by BhaRgav-MoRadiya** [comment]
+2

**Comment by moonlsd** [comment]
Same problem, HPA is completely broken when custom ENI is used after enabled secondary CIDR block for existing EKS cluster, we are in 1.18 already thus no workaround for us to force control plane restart.

**Comment by reedswenson** [comment]
+1

**Comment by mikestef9** [comment]
Hey everyone, there has a been a lot of confusion over the years on what exactly the subnets passed in during cluster creation are used for.

We have made updates to the [Cluster VPC considerations](https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html) EKS documentation page that attempts to clarify this topic. I'd also recommend looking at the [release notes](https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html#kubernetes-1.19) for EKS support Kubernetes version 1.19.

TLDR is that you can run worker nodes and load balancers in other subnets that are not registered with your EKS cluster. In fact, this is what we recommend. The subnets registered with an EKS cluster are only used for placing cross account network interfaces that facilitate Kubernetes exec/logs commands and private endpoint traffic flow. You can learn more about this architecture in this [blog](https://aws.amazon.com/blogs/containers/de-mystifying-cluster-networking-for-amazon-eks-worker-nodes/).

**Comment by cvvs** [comment]
Deploy a metrics-server to a non-registered subnet and you'll get 

`Message:               failing or missing response from https://100.127.108.27:4443/apis/metrics.k8s.io/v1beta1: Get https://100.127.108.27:4443/apis/metrics.k8s.io/v1beta1: Address is not allowed` 

messages in your logs.

**Comment by owengo** [comment]
We need to reconfigure  a subnet  ( ie destroy and create larger one with different properties) which is currently used by one of our eks cluster. This is not related to worker nodes ( which are using different subnets we can modify any time ). We just want to be able to reclaim a subnet without having to create a new cluster.

**Comment by thorro** [comment]
Don't do it. It ruined our cluster (worker nodes getting into NotReady state randomly). Worst of all, AWS allows you to do it without a warning. Just add a new subnet.

**Comment by nrvmodi** [comment]
In my case, I need internet-facing NLB.

The root cause in EKS is that you haven't selected a public subnet while creating the cluster.

After creating the cluster EKS will not allow to update subnets as of now here

To resolve the issue, I have performed the below steps

1. created a `public subnet` in the same vpc of EKS
2. Attached `IGW` in route tables in new created public subnets
3. Added below `tags` in public subnets
4. `kubernetes.io/cluster/<EKSClusterName>` : `shared`
Note: In a 4th step, Replace your EKS cluster name in placeholder `EKSClusterName`

**Comment by gbrancovici-sam** [comment]
Same issue here, please prioritize if possible.

**Comment by trucnguyenlam** [comment]
@mikestef9 would it be much better if there is a true documentation that shows step by step guide of how to make eks worker nodes available in the new subnets.

I find the current documentation is very hard to comprehend.

**Comment by donkeyx** [comment]
Is there any movement on an aws solution to this issue? 

It seems to have been hanging for a long time with lots of workarounds. This is an essential feature, considering the pods are using ips from the subnet allocation, which will eventually be exhausted as the cluster grows. So, next step would be to add some more subnets with additional ips, but in this case there is no simple way to add the required resources? 

This is a pretty big issue for a managed service to not be easily expandable.

**Comment by dogzzdogzz** [comment]
@donkeyx Just make sure all the routing and required tags for EKS are good in your new subnet and create ASG then, I expanded my IP pool by creating secondary subnets in each AZ for 20 clusters without any problem. I'm using self-managed ASG, i'm not sure if there is any configuration difference for managed node group.

Here is my terraform code for new subnet.
```terraform
locals {
    eks_clusters_tag = { for i in var.eks_clusters : "kubernetes.io/cluster/${i}" => "shared" if length(i) > 0 }
    eks_public_load_balancer_tag = length(var.eks_clusters) > 0 ? {"kubernetes.io/role/elb": "1"} : {}
    eks_internal_load_balancer_tag = length(var.eks_clusters) > 0 ? {"kubernetes.io/role/internal-elb": "1"} : {}
}

resource "aws_subnet" "secondary_private_subnet" {
  count             = var.enabled && var.secondary_private_subnet_enabled ? var.eks_cluster_az_count : 0
  vpc_id            = aws_vpc.mod.0.id

  cidr_block        = cidrsubnet(var.vpc_cidr, 3, count.index + local.az_count)
  availability_zone = element(data.aws_availability_zones.available.names, count.index)
  tags              = merge(
    module.default_label.tags, 
    map("Name", format("%s-private-secondary-%s", module.default_label.id, element(data.aws_availability_zones.available.names, count.index))), 
    local.eks_clusters_tag,
    local.eks_internal_load_balancer_tag,
  )
}

resource "aws_route_table" "secondary_private_subnet" {
  count            = var.enabled && var.secondary_private_subnet_enabled ? var.eks_cluster_az_count : 0
  vpc_id           = aws_vpc.mod.0.id
  tags             = merge(module.default_label.tags, map("Name", format("%s-private-secondary", module.default_label.id)))
}

resource "aws_route" "secondary_private_subnet_nat_gateway" {
  count                     = var.enabled && var.secondary_private_subnet_enabled ? var.eks_cluster_az_count : 0
  route_table_id            = element(aws_route_table.secondary_private_subnet.*.id, count.index)
  destination_cidr_block    = "0.0.0.0/0"
  nat_gateway_id            = element(aws_nat_gateway.natgw.*.id, count.index)
}

resource "aws_route_table_association" "secondary_private_subnet" {
  count          = var.enabled && var.secondary_private_subnet_enabled ? var.eks_cluster_az_count : 0
  subnet_id      = element(aws_subnet.secondary_private_subnet.*.id, count.index)
  route_table_id = element(aws_route_table.secondary_private_subnet.*.id, count.index)
}
```

**Comment by mikestef9** [comment]
@donkeyx see my comment above https://github.com/aws/containers-roadmap/issues/170#issuecomment-780174859

The subnets passed as part of cluster creation are not related to where you can run worker nodes.

**Comment by toabi** [comment]
Well, I think everybody knows now that putting workers in other subnets is no problem. The only issue which always stays like this is when you in the past did some error in subnet creation and made them in stupid sizes and now want to clean up.

Network layout cleanup is now still not possible without recreating the control plane.

**Comment by owengo** [comment]
@mikestef9 yes but it would be really useful to be able to reclaim the subnets passed as part of cluster creation. Right now the only solution is to destroy the cluster.

**Comment by sacredwx** [comment]
+1

**Comment by donkeyx** [comment]
@toabi and @owengo raise good points which match my scenario. 

In our case we had existing subnets whith a /24 range, which is pretty standard for our org.  But, only after creating and scaling the eks cluster, I noticed that we had exhaused out ips in the subnet. I then looked at how to use a custom range for the pods so it would have no impacton the subnet CIDR's, which, i think can only be done by "not" using managed node groups? The other solution, which is to add more subnets, would work but never release the subnets that were part of the original creation. 

Idealy, it would be nice for managed node groups to use an internal CIDR range so it does not impact existing subnets sizing. Alternative, though is a simple api or UI based option to "change" subnets, this is a managed service after all..

**Comment by hameno** [comment]
@donkeyx Managed nodes can already use separate subnets. I did exactly that. Deployed new managed node groups in new bigger subnets and everything works normally.

**Comment by diranged** [comment]
I'm a bit confused - it seems like a bunch of the conversation here is around the idea that you might not be able to add nodes to an EKS cluster if the subnets for those nodes are not part of the "control plane", but that's not true. My understanding is that the control plane just needs to be routable, not specifically in the same subnets.

We actually made the mistake early on to configure most of our EKS clusters so that the control planes are in "all" the subnets.. when in reality we only want them in 2-3 subnets. I am actually watching this ticket hoping that AWS will provide us a way to reconfigure a live EKS cluster to remove subnets that we do not need from the control plane. Is that something that is possible at all yet - even through a Support ticket?

**Comment by mnovacyu** [comment]
@mikestef9 do the subnets require specific tags of any sort? We've seen cases where the new secondary subnets are not getting recognized by the control plane.

**Comment by emaildanwilson** [comment]
I'm trying to decom a subnet which is a part of an eks config, it seems there is no way to do this other than by deleting the cluster. Is there a fix in progress for this?

**Comment by chrissnell-okta** [comment]
For what it's worth, toggling public/private had no effect on the recognizing of the new subnets.  Perhaps those who claim this works are just experiencing a coincidence of the reconciler running at the same time?  

Really wish there was a way to force a reconciler run.

**Comment by jamiegs** [comment]
So it seems that when your managed node group subnets and eks cluster subnets don't match you get an health issue error like this.   This makes it so I can't modify the node group any longer as any changes fails with this error.

```
AutoScalingGroupInvalidConfiguration	The Amazon AutoScalingGroup eks-node_workers has subnets ([subnet-....]) which is not expected by Amazon EKS. Expected subnets : ([subnet-...])
```

Is this a new health issue that was added? I don't see many references to it.

**Comment by armujahid** [comment]
I am also getting "AutoScalingGroupInvalidConfiguration The Amazon AutoScalingGroup eks-node_workers has 
subnets ([subnet-....]) which is not expected by Amazon EKS. Expected subnets : ([subnet-...])" issue after modifing ASG to use 3x private subnets instead of using 3x public subnets. Note that all 6 subnets were created during cluster creation.

**Comment by visla-xugeng** [comment]
My VPC has 3 public and 3 private subnets. I only select 3 private subnets, when I create EKS within this VPC. I created several managed node groups based these 3 private subnets. Then, I created a new managed node group in the public subnet in the EKS cluster. and It worked. I do not know how the eks can talk to the public subnet and place an managed node group in the public subnet. I thought I would get error, but did not. Can anyone explain?

**Comment by mikestef9** [comment]
@visla-xugeng see earlier comment https://github.com/aws/containers-roadmap/issues/170#issuecomment-780174859 from this thread

**Comment by visla-xugeng** [comment]
@mikestef9 
Thank you for your reply. I have read several doc mentioned in #170 comment. I just have one question here. Why do we still need to select several subnets when we create EKS cluster, since the nodes can be launched in any subnets in your cluster's VPC? What's the purpose of specifying subnets? 
In my case, I tagged both private and public subnets with ```kubernets.io/cluster/my-cluster shared```, when I created VPC. As a result, I can build node group in the both private and public subnets, and launch ec2 instances inside. It looks like that selecting several private subnets during the creation of EKS cluster is not necessary. Is this right?

**Comment by dz902** [comment]
This seems to be only a matter of UI problem. Why not just add it after all?

**Comment by sethatron** [comment]
My organization also really needs this functionality as well

**Comment by BEvgeniyS** [comment]
Hi @mikestef9,

Does the fact that this ticket moved back to 'Researching' mean we shouldn't expect any progress anytime soon?

**Comment by apetrosyan1613** [comment]
Hi all! I have I EKS in two subnets. Recently I added one public and one private subnets in a new AZ. I created a node pool in a private subnet and added NLB in public. Nodes were successfully created and pods were successfully scheduled. I can see pods logs and run kuberctl exec commands. kubectl top also outputs new nodes and pods. At first sight, everything looks good. What issues can I face?

**Comment by Sheepux** [comment]
Adding our vote/use case.
We're moving our cluster to private endpoint and at the same time moving the control plane endpoint outside the node subnet. However without being able to update we have to recreate the cluster itself and that's quite a struggle to explain to your users that our "maintenance" is going to be with the service being unavailable.
This issue is needed when doing some "small" architecture changes and mostly when you manage hundreds of cluster.

**Comment by pmyjavec** [comment]
So I'm having a similar experience. For us, we're now seeing the dreaded: `Address is not allowed` errors for some webhook calls after installing the "Calico" add-on exactly as outlined here: https://docs.aws.amazon.com/eks/latest/userguide/calico.html

Some of the pods which are having errors are now running in new subnets which are defined by calico [IPPool resources](https://projectcalico.docs.tigera.io/reference/resources/ippool), what do we do here? @mikestef9 you mentioned there is some type of reconciler running on the control plane. Is there some way I can have it pickup the new subnets defined in calico? 

I'll raise a support ticket with AWS as well because I think without having this functionality, the "calico add-on" is fairly worthless.

**Comment by florenp** [comment]
@mikestef9 How could we be sure that the proxy-cidr-whitelist flags is updated with the new CIDR?
Is there a way to check the value?

**Comment by evercast-mahesh2021** [comment]
Any updates on this? We are also facing subnets are running out of available IP issue. I have tried the workaround, created subnets with correct tags, added them to route table and NACL etc. Also, i added them to ASG group. I see the new instance are coming up but not recognized by EKS cluster. We can't recreate the cluster for just adding new subnets at this time. Let me know if there is any other workaround or solution on this.

**Comment by jku8** [comment]
@florenp you can try enabling API server logs in control plane logging: https://docs.aws.amazon.com/eks/latest/userguide/control-plane-logs.html

**Comment by jukie** [comment]
Adding another bump.
Recently we hit an incident after a change in how subnets are returned (two subnets covering the same AZ) which brought to light a bad config on our end. However now we're stuck because we can't easily remove the problematic subnet.

**Comment by revawiki** [comment]
Vote for this feature..
it gives flexibility on when we move the control plane to private subnet.
Really useful in practice..

**Comment by mikestef9** [comment]
**What's new**: https://aws.amazon.com/about-aws/whats-new/2023/10/amazon-eks-modification-cluster-subnets-security/
**Docs**: https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html

**Important Note**: The subnets passed as an update must contain the same set of AZs as originally passed when creating the cluster.

**Comment by diranged** [comment]
@GoodMirek,
  This is great news.. any idea how long until CloudFormation supports this?

**Comment by mikestef9** [comment]
@diranged CFN support is already available. See [docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-eks-cluster.html#cfn-eks-cluster-resourcesvpcconfig) changed to "No Interruption"

**Comment by mikestef9** [comment]
@GoodMirek We have added a note to docs

https://docs.aws.amazon.com/eks/latest/APIReference/API_UpdateClusterConfig.html

**Comment by cuppett** [comment]
@mikestef9 On "[Creating an cluster](https://docs.aws.amazon.com/eks/latest/userguide/create-cluster.html)", it still has language like this (inside eksctl method):

> The subnets that you choose must meet the [Amazon EKS subnet requirements](https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html#network-requirements-subnets). Before selecting subnets, we recommend that you're familiar with all of the [Amazon EKS VPC and subnet requirements and considerations](https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html). **You can't change which subnets you want to use after cluster creation.**

Can probably just remove those sentences wherever they might appear.

**Comment by somdoron** [comment]
It seems that adding new availability zones is still not possible with this change. Does an issue exist to track that?

**Comment by mikestef9** [comment]
> It seems that adding new availability zones is still not possible with this change. Does an issue exist to track that?

You'd need to create a new issue with that request. It would be good if you elaborate on the use case in a new issue as well. To re-iterate, the control plane subnets have nothing to do with worker nodes, you can run worker nodes in subnets in different AZs than the ones associated the control plane.

**Comment by mikestef9** [comment]
Deep dive blog published

https://aws.amazon.com/blogs/containers/enhanced-vpc-flexibility-modify-subnets-and-security-groups-in-amazon-eks/

**Comment by mikestef9** [comment]
The Kubernetes control plane managed by EKS is always spread across 3 AZs, customers can't control that. The subnets passed to EKS only influence where EKS places cross account ENIs in your VPC to support exec/logs and the private endpoint. You can run worker nodes in any subnets/AZs you want in the same VPC, including subnets not passed to EKS.

**Comment by bryantbiggs** [comment]
> It would make sense to have endpoints in each AZ for the control plane where worker nodes also reside.

FYI - even if you were to supply more subnets in additional AZs to the control plane, there would still only be 2 ENIs created by EKS. EKS does not provision an ENI in each subnet that is provided to the control plane. 

Also, you can always expand the number of subnets/AZs used by the nodes in your data plane (supplied to your node groups) without needing to modify the control plane. Provided your VPC is setup to route traffic (there is a path for pods to reach the control plane and vice versa) between the subnets used by the control plane ENIs and the subnets used by the nodes, you can modify this as you see fit and expand the number of AZs used by your data plane.

**Comment by lkoniecz** [comment]
> > The Kubernetes control plane managed by EKS is always spread across 3 AZs, customers can't control that. The subnets passed to EKS only influence where EKS places cross account ENIs in your VPC to support exec/logs and the private endpoint. You can run worker nodes in any subnets/AZs you want in the same VPC, including subnets not passed to EKS.
> 
> The issue here @mikestef9 is that an existing cluster, where these cross-account ENIs were only created in two availability zones originally still cannot be extended to three or more AZs in future. It would make sense to have endpoints in each AZ for the control plane where worker nodes also reside. This is currently still not supported and must be a FR for enhancement.

Anyone aware of such FR being opened?

We have several clusters initially created across three Availability Zones (AZs). Despite having worker nodes in only two AZs, we still incur unnecessary costs for inter-AZ traffic between the worker nodes and the third AZ where the EKS cluster's ENI might be provisioned.
